### PR TITLE
feat(forms): make reactive forms strongly typed 

### PR DESCRIPTION
**Pull request by KostyaTretyak**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Feature

## What is the current behavior?

Weak typechecking for reactive forms.

Issue Number: #13721

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: packages/forms/src/form_builder.ts [hunk 1]
```diff
@@ -18,6 +18,18 @@ function isAbstractControlOptions(options: AbstractControlOptions|
...
+ * Form builder control config.
+ */
+export type FbControlConfig<T> =
+  | AbstractControl<T>
+  | (T | { value: T; disabled: boolean })
+  | [
+      T | { value: T; disabled: boolean },
+      (ValidatorFn | ValidatorFn[] | AbstractControlOptions)?,
+      (AsyncValidatorFn | AsyncValidatorFn[])?
+    ];
```
Comment (SchnWalter) [inline packages/forms/src/form_builder.ts:31]: 1) I would avoid adding the `Fb` prefix, maybe it should be `FormControlConfig<T>`. These are basically the individual constructor arguments for the `FormControl`, `FormGroup` and `FormArray` classes; all of which extend `AbstractControl`, and these exact arguments are also used by the form builder.

2) For strict typing and backward compatibility, It should accept `null` and `undefined` as value. But I think that here we could improve the code a bit. If it were up to me, I would try to get rid of `null` in favor of only using `undefined`. (`typeof null === 'object')

2) You should introduce some extra types, this type is intimidating at a first look. Maybe it could look something like this:

```ts
// Maybe FormState<T> should also include undefined.
// @see AbstractControl._isBoxedValue()
// @see FormControl._applyFormState()
// @see FormBuilder.control()
export type FormState<T> = null | T | {
  value: T;
  disabled: boolean;
}

// @see coerceToValidator()
export type FormValidatorOrOptions = undefined | null | ValidatorFn|ValidatorFn[]|AbstractControlOptions;

// @see coerceToAsyncValidator()
export type FormAsyncValidator = undefined | null | AsyncValidatorFn|AsyncValidatorFn[];

export type FormControlConfig<T> = FormState<T> | FormValidatorOrOptions |  FormAsyncValidator;
```

If you add these extra types, we can start using them in other methods.

**Review by SchnWalter — CHANGES_REQUESTED** [review]

File: packages/forms/src/form_builder.ts (hunk 1 above)
Comment (KostyaTretyak) [reply packages/forms/src/form_builder.ts:31]: 1. agree
2. agree
3. need to think

**Comment by KostyaTretyak** [comment]
@AndrewKushnir, please tell me why you set "flag: breaking change"?

**Comment by KostyaTretyak** [comment]
Good news! After sleep, I see why "flag: breaking change" is seted here =). This is not a runtime breaking change, but only by type.

**Comment by AndrewKushnir** [comment]
Hi @KostyaTretyak, thanks for creating this PR! We will review it and provide the feedback.

> please tell me why you set "flag: breaking change"?

I've added the "flag: breaking change" flag since there are changes to the public API that may not be backwards-compatible. For example [this change](https://github.com/angular/angular/pull/37389/files#diff-45964408601d1f145626eafb6c11a9a3L204) to the `status` field type from `string` -> enum with the set of specific values looks potentially backwards-incompatible (there might be more cases, I can provide additional info after detailed review). The "breaking change" flag is an indication that the changes should be processed in a certain way (special release notes, running a full test suite in Google codebase, etc) and be included into the major Angular versions.

Thank you.

**Comment by KostyaTretyak** [comment]
I can't find in the repo tests that fail. Can someone help me with this?

I try to follow the instructions at [running-tests-locally](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#running-tests-locally), but can't interpret error output in `*.log` files.

**Comment by KostyaTretyak** [comment]
After running

```bash
yarn bazel test //... --build_tag_filters=-ivy-only --test_tag_filters=-ivy-only
```
I have

```text
ERROR: /home/me/.cache/bazel/_bazel_me/acdbb7898fe4159d08f1a2ee035c90a5/external/com_google_protobuf/BUILD:267:1: C++ compilation of rule '@com_google_protobuf//:protoc_lib' failed (Exit 1) gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 32 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
gcc: error trying to exec 'cc1plus': execvp: No such file or directory
```

**Comment by KostyaTretyak** [comment]
I trying follow the instructions at [DEVELOPER](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md).
- I have git v2.17.1, node v12.16.3, yarn v1.22.4 and Java Development Kit v11.0.7
- On step "Buidling" `node ./scripts/build/build-packages-dist.js` I have:

```text
ERROR: /home/me/.cache/bazel/_bazel_me/acdbb7898fe4159d08f1a2ee035c90a5/external/com_google_protobuf/BUILD:267:1: C++ compilation of rule '@com_google_protobuf//:protoc_lib' failed (Exit 1) gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 32 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
gcc: error trying to exec 'cc1plus': execvp: No such file or directory
INFO: Elapsed time: 19.945s, Critical Path: 7.64s
INFO: 8 processes: 8 linux-sandbox.
FAILED: Build did NOT complete successfully
child_process.js:669
    throw err;
    ^

Error: Command failed: yarn --cwd "/srv/git/angular/angular" --silent bazel build --config=release --config=view-engine //packages/upgrade:npm_package //packages/service-worker:npm_package //packages/router:npm_package //packages/platform-webworker-dynamic:npm_package //packages/platform-webworker:npm_package //packages/platform-server:npm_package //packages/platform-browser-dynamic:npm_package //packages/platform-browser:npm_package //packages/localize:npm_package //packages/language-service:npm_package //packages/forms:npm_package //packages/elements:npm_package //packages/core:npm_package //packages/compiler-cli:npm_package //packages/compiler:npm_package //packages/common:npm_package //packages/bazel:npm_package //packages/animations:npm_package
    at checkExecSyncError (child_process.js:630:11)
    at execSync (child_process.js:666:15)
    at exec (/srv/git/angular/angular/scripts/build/package-builder.js:116:18)
    at buildTargetPackages (/srv/git/angular/angular/scripts/build/package-builder.js:76:3)
    at Object.<anonymous> (/srv/git/angular/angular/scripts/build/build-packages-dist.js:18:1)
    at Module._compile (internal/modules/cjs/loader.js:1133:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1153:10)
    at Module.load (internal/modules/cjs/loader.js:977:32)
    at Function.Module._load (internal/modules/cjs/loader.js:877:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12) {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 14600,
  stdout: null,
  stderr: null
}
```

I can't figure out what mean that six errors from CI. For example, take the [first error](https://circleci.com/gh/angular/angular/725631?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link). There I see the following error:

```text
ERROR: /tmp/angular-components-repo/src/material/select/testing/BUILD.bazel:25:1: Compiling TypeScript (devmode) //src/material/select/testing:harness_tests_lib failed (Exit 1)
src/material/select/testing/shared.spec.ts:234:32 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(expected: Expected<null | undefined>, expectationFailOutput?: any): boolean', gave the following error.
    Argument of type '"CA"' is not assignable to parameter of type 'Expected<null | undefined>'.
  Overload 2 of 2, '(expected: Expected<null | undefined>, expectationFailOutput?: any): boolean', gave the following error.
    Argument of type '"CA"' is not assignable to parameter of type 'Expected<null | undefined>'.

234     expect(control.value).toBe('CA');
```

How I can fix this? Can anyone help me with this?

File: packages/forms/test/value_accessor_integration_spec.ts [hunk 2]
```diff
@@ -663,8 +663,8 @@ import {dispatchEvent} from '@angular/platform-browser/testing/src/browser_util'
         it('should support removing controls from <type=radio>', () => {
           const fixture = initTest(FormControlRadioButtons);
           const showRadio = new FormControl('yes');
-          const form =
-              new FormGroup({'food': new FormControl('fish'), 'drink': new FormControl('sprite')});
+          const form = new FormGroup<any>(
+              {'food': new FormControl('fish'), 'drink': new FormControl('sprite')});
```
Comment (SchnWalter) [inline packages/forms/test/value_accessor_integration_spec.ts:667]: No need to add `any` in here, we already have the type in the class definition: `<T extends object = any>` and it's properly reused in the constructor to pass in the individual types to the all the `AbstractControl` instances.

But since here we are changing the values, maybe we should just provide the interface and use it as a `Partial<X>`:
```ts
interface TestFormData {
  food: string; drink: string;
}
const form = new FormGroup<Partial<TestFormData>>( /* ... */ );
```
This way, you are always forced to check if your control exists.

File: packages/forms/test/reactive_integration_spec.ts [hunk 3]
```diff
@@ -1403,7 +1405,7 @@ import {MyInput, MyInputForm} from './value_accessor_integration_spec';
           const fixture = initTest(FormGroupComp);
           const control =
               new FormControl('', {validators: Validators.required, updateOn: 'submit'});
-          const formGroup = new FormGroup({login: control});
+          const formGroup = new FormGroup<any>({login: control});
```
Comment (SchnWalter) [inline packages/forms/test/reactive_integration_spec.ts:1408]: The `any` type should be avoided in these tests. We want to also make sure that your types are propagated properly. Here, we should have a local interface:
```
interface TestFormGroupData { login:  string };
```

File: packages/forms/src/model.ts [hunk 4]
```diff
@@ -201,7 +214,7 @@ export abstract class AbstractControl {
    * both valid AND invalid or invalid AND disabled.
    */
   // TODO(issue/24571): remove '!'.
-  public readonly status!: string;
+  public readonly status!: FormStatus;
```
Comment (SchnWalter) [inline packages/forms/src/model.ts:217]: This should be made backward compatible in some way. Also, I would avoid changing the status in the current PR, I would stick to adding types to the form controls.

File: packages/forms/test/value_accessor_integration_spec.ts (hunk 2 above)
Comment (KostyaTretyak) [reply packages/forms/test/value_accessor_integration_spec.ts:667]: Please note that I did write my own library for typing angular forms. I understand such basic things in TypeScript.

It was possible to do `Partial<TestFormData>`, but at least here it is necessary to specify `any`. TypeScript takes into account the default value for generics `<T extends object = any>`, but it has a lower priority than the type of data passed to the constructor.

**Comment by KostyaTretyak** [comment]
1. I follow [Submitting a Pull Request (PR)](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-submitting-a-pull-request-pr) and [Running Tests Locally](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#running-tests-locally).
2. I found that [ci/circleci: components-repo-unit-tests](https://circleci.com/gh/angular/angular/727075?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) I can't fix because this error is about [angular/components repo](https://github.com/angular/components/blob/72c38f7514faa7da2c3674c9a571c5f1693e7990/src/components-examples/material/form-field/form-field-theming/form-field-theming-example.ts#L23)
3. It seems that the following errors are actually the same error (with same commant `yarn --cwd aio e2e --configuration=ci`):
- [ci/circleci: test_aio](https://circleci.com/gh/angular/angular/727304?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)
- [ci/circleci: test_aio_local](https://circleci.com/gh/angular/angular/727316?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)
- [ci/circleci: test_aio_local_viewengine](https://circleci.com/gh/angular/angular/727310?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

I tried to run `yarn --cwd aio e2e --configuration = ci` and` yarn --cwd aio e2e`, but my tests failed with the following error:

```text
yarn run v1.22.4
$ yarn check-env && yarn update-webdriver
$ yarn ~~check-env
$ node scripts/check-environment
$ yarn aio-check-local
$ node tools/ng-packages-installer check .
$ node ../scripts/webdriver-manager-update.js
error Command "webdriver-manager" not found.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[webdriver-manager-update.js] Call to 'yarn webdriver-manager update --gecko=false --standalone=false --versions.chrome 80.0.3987.0' failed with error code 1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

If anyone knows how to fix this, please tell me how.

File: packages/forms/test/value_accessor_integration_spec.ts (hunk 2 above)
Comment (SchnWalter) [reply packages/forms/test/value_accessor_integration_spec.ts:667]: Maybe, instead of
```ts
export declare class FormBuilder {
  group<T extends object = any<( /* ... */ );
  // ...
}
```
we could use:
```ts
export declare class FormBuilder {
  group<T extends Record<string, any> = Record<string, any>>( /* ... */ )
  // ...
}
```

P.S. I didn't mean to imply anything. Sorry for offending you.

Comment (KostyaTretyak) [reply packages/forms/test/value_accessor_integration_spec.ts:667]: What issues can there be with `<T extends object = any>`?

File: packages/forms/test/form_array_spec.ts [hunk 5]
```diff
@@ -115,7 +115,7 @@ describe('FormArray', () => {
 
   describe('markAllAsTouched', () => {
     it('should mark all descendants as touched', () => {
-      const formArray: FormArray = new FormArray([
+      const formArray: FormArray = new FormArray<string|{c1: string}|{c2: string}[]>([
```
Comment (SchnWalter) [inline packages/forms/test/form_array_spec.ts:118]: You should avoid using the `[]` notation on complex types, it is a source of confusion and errors. You should replace this with:
```
string|{c1: string}|Array<{c2: string}>
```

The [array-type](https://palantir.github.io/tslint/rules/array-type/) tslint rule would be helpful here.

File: packages/forms/test/reactive_integration_spec.ts [hunk 6]
```diff
@@ -1357,7 +1359,7 @@ import {MyInput, MyInputForm} from './value_accessor_integration_spec';
           fixture.detectChanges();
 
           expect(formGroup.value)
-              .toEqual({login: null}, 'Expected form value to stay empty on submit');
+              .toEqual({login: null} as any, 'Expected form value to stay empty on submit');
```
Comment (SchnWalter) [inline packages/forms/test/reactive_integration_spec.ts:1362]: As pointed in my initial comment, the controls also accept null and undefined values. This type cast hides the fact that the types don't accept null as a value.

Comment (KostyaTretyak) [reply packages/forms/test/reactive_integration_spec.ts:1362]: This type is not about constructor of a form control (as you pointed in your initial comment), it's about form value.

File: packages/forms/test/value_accessor_integration_spec.ts (hunk 2 above)
Comment (SchnWalter) [reply packages/forms/test/value_accessor_integration_spec.ts:667]: I was hoping to prevent accidental passing of arrays, but it's still allowed, even with `extends Record<X>` (I never knew that)

```ts

type ArrTest = string[];

interface ITest {
  a: string;
}

type FormStateExtendsObject<T extends object = any> = T;
const obj1x: FormStateExtendsObject = { a: '1' };
const obj2x: FormStateExtendsObject = ['1'];
const obj1y: FormStateExtendsObject<ITest> = { a: '1' };
const obj2y: FormStateExtendsObject<ArrTest> = ['1'];

type FormStateExtendsRecord<T extends Record<any, any> = Record<any, any>> = T;
const rec1x: FormStateExtendsRecord = { a: '1' };
const rec2x: FormStateExtendsRecord = ['1'];  // wanted to see an error here.
const rec1y: FormStateExtendsRecord<ITest> = { a: '1' };
const rec2y: FormStateExtendsRecord<ArrTest> = ['1'];  // wanted to see an error here.

```

File: packages/forms/test/form_array_spec.ts (hunk 5 above)
Comment (KostyaTretyak) [reply packages/forms/test/form_array_spec.ts:118]: Here no array-type tslint rule, hence I can use and `[]`.

**Comment by KostyaTretyak** [comment]
It remains to fix one error. This error, as far as I understand, should be corrected on the [angular/components repo](https://github.com/angular/components/blob/72c38f7514faa7da2c3674c9a571c5f1693e7990/src/components-examples/material/form-field/form-field-theming/form-field-theming-example.ts#L23).

**Comment by jonkoops** [comment]
@KostyaTretyak If you need any help with this PR I am willing to help out. Perhaps my PR for supporting a generic type for `ControlValueAccessor` can be incorporated here? https://github.com/angular/angular/pull/31801/files

**Comment by KostyaTretyak** [comment]
> Perhaps my PR for supporting a generic type for ControlValueAccessor can be incorporated here (see: #31801)?

@jonkoops, I still want to add form status type and form validation type to the form module. But in the following PR, not in current PR. Although, if the Angular team says it's better to do it in the current PR, then I'm ready to do it. The same goes for `ControlValueAccessor`.

File: packages/forms/src/form_builder.ts (hunk 1 above)
Comment (SchnWalter) [reply packages/forms/src/form_builder.ts:31]: Nice, just adding that `FormState` made this a lot more readable. Thanks!

(I'll give it a spin now)

**Comment by SchnWalter** [comment]
> I tried to run `yarn --cwd aio e2e --configuration = ci` and` yarn --cwd aio e2e`, but my tests failed with the following error:
> ...
> error Command "webdriver-manager" not found.

@KostyaTretyak, it looks like you also need to run `yarn install` in the `aio` directory to get around that error. `yarn install` in the root directory doesn't seem to be sufficient.

```bash
# start with a clean slate, if you want to.
# rm -rf aio/node_modules
yarn --cwd aio install
```
But then you run into other issues, so you need to also run:
```
yarn --cwd aio setup-local
```
And only the, the `yarn --cwd aio e2e --configuration=ci` passes.

Checkout the "Developer tasks" section in [aio/README.md](https://github.com/angular/angular/blob/master/aio/README.md#developer-tasks). Maybe these `aio` specific things should also be mentioned somewhere in the "Running Tests Locally" section of [docs/DEVELOPER.md](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md)`, it would be nice to just follow that section and have all tests running locally. You can follow #33591 to get updates when the docs are going to be updated. cc: @aikidave

**Comment by AndrewKushnir** [comment]
Thanks for the updates in this PR @KostyaTretyak and thanks @SchnWalter for providing feedback
👍 

As Kara mentioned in https://github.com/angular/angular/issues/13721#issuecomment-637698836, we'll start looking into Form type improvements soon (we don't have an exact ETA at this moment). Meanwhile I've collected all types-related PRs in the [Forms type improvements](https://github.com/angular/angular/milestone/103) Milestone on Github, so we can take them (including comments) into account.

> I still want to add form status type and form validation type to the form module. But in the following PR, not in current PR. Although, if the Angular team says it's better to do it in the current PR, then I'm ready to do it. The same goes for ControlValueAccessor.

I think it makes sense to have multiple individual PRs targeting subsets of Forms package, so it's easier to review, test and merge (in master branch as well as in Google codebase internally).

We'll provide more updates as soon as we have them.

Thank you.

File: goldens/public-api/forms/forms.d.ts [hunk 7]
```diff
@@ -265,32 +269,35 @@ export declare class FormControlName extends NgControl implements OnChanges, OnD
     viewToModelUpdate(newValue: any): void;
 }
 
-export declare class FormGroup extends AbstractControl {
+export declare type FormControlState<T> = null | T | {
+    value: null | T;
+    disabled: boolean;
+};
+
+export declare class FormGroup<T extends object = any> extends AbstractControl<T> {
```
Comment (gaplo917) [inline goldens/public-api/forms/forms.d.ts:277]: If `FormGroup` is going to use value type for `T`, the implementation will lose the ControlType and can not chained to access the nested control

Consider there is a nested FormGroup, you cannot type-safely do the following
```
let formGroup: FormGroup<{ name: string, address: { ad1: string ad2: string } }>

formGroup.controls.address.controls.ad1 // this will not compile with the Typescripts, as `.control.address` is AbstractControl<string>
```

Comment (KostyaTretyak) [reply goldens/public-api/forms/forms.d.ts:277]: Yes, it is. To fix this you need:

```ts
declare let formGroup: FormGroup<{ name: string, address: { ad1: string, ad2: string } }>;
(formGroup.controls.address as FormGroup).controls.ad1
```

In my own module [@ng-stack/forms](https://github.com/KostyaTretyak/ng-stack/tree/master/projects/forms#automatically-detect-appropriate-types-for-form-controls) this is done automatically based on the data type of the model property. But, unfortunately, this mechanism does not work perfectly because the data type `object` can be assigned to both `FormGroup` and `FormControl`. As a workaround, I created a new type `Control<T>` to give a TypeScript hint that the property should contain `FormControl`. But I doubt that the Angular team will want to implement `Control<T>`.

Comment (gaplo917) [reply goldens/public-api/forms/forms.d.ts:277]: @KostyaTretyak I have faced this issues last few months and solved this issues by using `infer` type. You could check my implementation of `TypedFormGroup`.

`TypedFormGroup` has no api changes, just extends the original `FormGroup` and add typescript typing.

TypedFormGroup: https://github.com/gaplo917/angular-typed-forms/blob/master/projects/angular-typed-forms/src/lib/forms/typed-form-group.ts

Infer Type: https://github.com/gaplo917/angular-typed-forms/blob/master/projects/angular-typed-forms/src/lib/types/index.ts

You could clone to test on the Playground Repo (CodeSandBox still not working)
https://github.com/gaplo917/angular-typed-forms-playground

My implementation achieve 100% strict type check on complex form group `controls` and also the `setValue` & `patchValue` api. As the above implementation has no changes of the FormGroup API, the migration to Typed FormGroup is completely smooth.

Comment (KostyaTretyak) [reply goldens/public-api/forms/forms.d.ts:277]: @gaplo917, you need to fix your `TypedFormBuilder` API, because this is not works:

```ts
interface Model {
  name: TypedFormControl<string>,
  address: TypedFormGroup<{ ad1: TypedFormControl<string>, ad2: TypedFormControl<string> }>
}

const fb = new TypedFormBuilder();

const formGroup = fb.group<Model>({
  name: '',
  address: fb.group({
    ad1: '',
    ad2: '',
  })
});
```

Comment (KostyaTretyak) [reply goldens/public-api/forms/forms.d.ts:277]: Here are a couple of problems:
1. If we follow your approach, we need to create two different models - for form controls and for form values.
2. The model for form controls is harder to read.

Comment (KostyaTretyak) [reply goldens/public-api/forms/forms.d.ts:277]: You need to fix your `TypedFormGroup` API, because this is not throw TypeScript error, but it's wrong:

```ts
let formGroup: TypedFormGroup<string>;
```

Comment (KostyaTretyak) [reply goldens/public-api/forms/forms.d.ts:277]: You need to fix your `TypedFormArray` API, because this is not works:

```ts
const formArray = new TypedFormArray([
  new TypedFormControl('')
]);
```

Comment (gaplo917) [reply goldens/public-api/forms/forms.d.ts:277]: @KostyaTretyak Thanks for pointing out

> @gaplo917, you need to fix your `TypedFormBuilder` API, because this is not works:
> 
> ```ts
> interface Model {
>   name: TypedFormControl<string>,
>   address: TypedFormGroup<{ ad1: TypedFormControl<string>, ad2: TypedFormControl<string> }>
> }
> 
> const fb = new TypedFormBuilder();
> 
> const formGroup = fb.group<Model>({
>   name: '',
>   address: fb.group({
>     ad1: '',
>     ad2: '',
>   })
> });
> ```

Under the current design, it should be written in this way.
```
interface Model {
  name: TypedFormControl<string>,
  address: TypedFormGroup<{ ad1: TypedFormControl<string>, ad2: TypedFormControl<string> }>
}

const fb = new TypedFormBuilder();

const formGroup = fb.group<Model>({
  name: fb.control(''),
  address: fb.group({
    ad1: fb.control(''),
    ad2: fb.control(''),
  })
});
```

Comment (gaplo917) [reply goldens/public-api/forms/forms.d.ts:277]: > You need to fix your `TypedFormGroup` API, because this is not throw TypeScript error, but it's wrong:
> 
> ```ts
> let formGroup: TypedFormGroup<string>;
> ```

Thanks for pointing out. Maybe it is a typescript issues.

By the generic type constraint definition,
```ts
export declare type KeyValueControl<T> = {
    [key in keyof T]: T[key] & AbstractControl;
};

class TypedFormGroup<T extends KeyValueControl<T>> extends FormGroup {}
```

`string` is clearly not extends `{
  [key in keyof T]: T[key] & AbstractControl;
}`, it should be throw compile error by typescript compiler

Comment (gaplo917) [reply goldens/public-api/forms/forms.d.ts:277]: > You need to fix your `TypedFormArray` API, because this is not works:
> 
> ```ts
> const formArray = new TypedFormArray([
>   new TypedFormControl('')
> ]);
> ```

This is definitely a problem. I broken the compatibility of the origin API. Thanks for pointing out!

Comment (gaplo917) [reply goldens/public-api/forms/forms.d.ts:277]: All issues you have mentioned are solved in 2.0.0 version. I finally made a CodeSandBox-ready version that can easily to fork, change and test.

https://codesandbox.io/s/github/gaplo917/angular-typed-form-codesandbox?file=/src/app/demo/demo.component.ts

Thanks for pointing out the edge cases again. 👍

> Here are a couple of problems:
> 
>     1. If we follow your approach, we need to create two different models - for form controls and for form values.


My points is to completely rely on typescript type inference. In real world application, most of the time DON'T need to know the form `ValueType` once you have a `ControlType`. 

Consider the common scenario: 
1. You get data from remote API (you already have a API data model)
2. The API data model most likely can not be directly used as a form structure, you very likely need to apply a transformation (You have to create a new ValueType)
3. When you submit the form,  you also very likely to apply a transformation to a standard API request body (Another new type).

```ts
interface FooApiData {
   id: string
   age: number
   dob: string | null
   createdAt: string
}

interface FooFormType {
  id: TypedFormControl<string>
  age: TypedFormControl<number>
  dob: TypedFormControl<Date | null>
} 

interface FooApiUpdateRequest {
   id: string
   dob: string
   age: number
}

class FooForm extends TypedFormGroup<FooFormType> {
    constructor(private fb: TypedFormBuilder, private initialValue: FooApiData) {
      super({
          id: fb.control(initialValue.id, Validators.required),
          dob: fb.control(initialValue.dob === null ? new Date(initialValue.dob) : null),
          age: fb.number(initialValue.age, Validators.required)
       })
   }
   toRequestBody(): FooApiUpdateRequest {
       const typedValue = this.value
       return {
          id: value.id,
          dob: value.dob !== null ? moment(value.dob).format('YYYYMMDD') : null,
          age: value.number
       }
   }
}

const apiData: FooApiData = apiService.getFoo()
const form: FooForm = new FooForm(new TypedFormBuilder())
form.setValue({
  id: apiData.id,
  age: apiData.age
})

// assume some UI changes the form value
if(form.valid){
  const payload = form.toRequestBody()
  apiService.updateFoo(payload)
}

```

In my point of view, using `ValueType` as `T` also have problems, 
1. You cannot have a strict type check on the Angular HTML if you have a complex form structure. i.e.  `FormArray<{ id: string, age: number} >`, `FormGroup<{ id: string, address: [{ ad1: string, ad2: string }]>` 
2. You need to create many fragmented types or `Pick<T>/Omit<T>/Record<T>...` when doing type cast. This would happen in every place when the form structure is complex.
3.  After you do a force type cast. Changing a `ValueType` fields type will NOT trigger a compile type-mismatch error on the subsequence type cast. You need to manual check everytime.

> 
>     2. The model for form controls is harder to read.

Our production application has 4 levels combined of FormGroup or FormArray, our junior colleagues still can follow it accurately. (Most of the time reply on the typescript auto complete features and protected by typedscript compiler and Angular `fullTemplateTypeCheck` feature)

I foresee a manually a type cast randomly when needed would be a disaster in our application because it does NOT propagated the type changes when changing the Root `ValueType`.
```
```

Comment (KostyaTretyak) [reply goldens/public-api/forms/forms.d.ts:277]: @gaplo917, try `@ng-stack/forms`, this module works without need type cast. Gradually, I plan to implements all the features of `@ng-stack/forms` to Angular. This I just want to do in a few pull requests.

I also use `infer`, and for a long time. See [ControlType](https://github.com/KostyaTretyak/ng-stack/blob/forms%402.0.1/projects/forms/src/lib/types.ts#L88-L96)

Comment (gaplo917) [reply goldens/public-api/forms/forms.d.ts:277]: @KostyaTretyak  Thanks for your sharing.

One possible draw back I have found in your implementation. (Correct me if I am wrong) As `@ng-stack/forms`  always compose and extract with a base `FormGroup / FormArray` type. 

https://github.com/KostyaTretyak/ng-stack/blob/d784d7870cf84a9216327e75ffd97fef45f5c7ce/projects/forms/src/lib/types.ts#L89

There is no way to encapsulate some complex logic by dividing a sub-forms because developers cannot easily make a abstraction then divide the logic to many sub-forms unless they have to spend a lots of time to build their own `infer` type.

For example,
```ts
interface FooFormData {
  id: string
  age: number
  dob: Date | null
} 
interface RootFormData {
   foo: FooFormData
}

// `@ng-stack/forms` cannot do the followings

class FooForm extends FormGroup<FooFormData> {
   toRequestBody() {..}
}
class RootForm extends FormGroup<RootFormData> {
    toRequestBody() {
       return {
          foo: this.controls.foo.toRequestBody()
       }
    }
}

// Or something like
const fooReqBody = fb.group({...}).controls.foo.toRequestBody()
```
Any way to solve this in `@ng-stack/forms`?

Sorry for bothering. Just want to know the difference and limitation of the approaches.

Comment (KostyaTretyak) [reply goldens/public-api/forms/forms.d.ts:277]: I have never had to implement many `toRequestBody()` methods for single HTTP request in my practice. I always send `form.value` and see no problems with it.

Comment (KostyaTretyak) [reply goldens/public-api/forms/forms.d.ts:277]: By the way, you can always make different transformations for different models:

```ts
class Model1 {
  static toRequestBody() {...}
}

class Model2 {
  static toRequestBody() {...}
}

const requestBody1 = Model1.toRequestBody(form1.value);
const requestBody2 = Model2.toRequestBody(form2.value);
```

**Comment by jayoungers** [comment]
Due to the HTML binding issues noted in issue #36405, it would be nice if the typings on `FormGroup` and `FormArray` included the type of control itself.

I.e. instead of:

FormArray - `controls: AbstractControl<Item>[];`
FormGroup - `controls: { [key in keyof T]: AbstractControl<T[key]>; };`

The generics were something like: `<Item, Control extends AbstractControl<Item>>`

If possible I think that would resolve a lot of work-arounds currently needed, especially since it appears angular 10 is more aggressive with `strictTemplates`

**Comment by KostyaTretyak** [comment]
My opinion on this can be read in [this discussion](https://github.com/angular/angular/issues/13721#issuecomment-654223116).

**Comment by AndrewKushnir** [comment]
Hi @KostyaTretyak,

Quick update: while this task is in our Backlog at this moment (i.e. we have not started active work on it) I ran tests in Google's codebase with the changes from this PR. As we expected, it's breaking a relatively big number of apps, so landing this PR with the current set of changes would be difficult (since that would require adapting apps code to this change). This also indicates that landing this PR on NPM (as a part of a major release) would require Forms users to adapt their apps during the upgrade as well. This is expected for some breaking changes (where creating a schematic is not feasible or a breaking change is affecting a small portion of the codebase and the fix is straightforward), but we always try to minimize the impact of breaking changes to provide a smooth upgrade experience.

The changes in this PR are very useful and I'd like to propose splitting this PR into a set of smaller PRs and try to land them in the following sequence (each group might have more than one PR):

- Backwards-compatible changes. These changes should not require any work from apps authors during the upgrade.
- Breaking changes but with limited impact and straightforward fix(es). Good example of these changes can be found in PR #32671.
- Breaking changes with high impact affecting a large portion of the codebase and/or where the fix is not straightforward. For these changes we should perform an assessment to see how we can land them best to ensure smooth upgrades.

Please let me know your opinion on how the changes in this PR can be split and whether we can extract some backwards-compatible changes into separate PR(s).

Thank you.

**Comment by KostyaTretyak** [comment]
Unfortunately, I will not be able to take the time to improve this Pull Request in the near future.
